File: frontend/lib/services/complaint_service.dart
import '../models/complaint_model.dart';
import 'api_service.dart';

class ComplaintService {
  final ApiService _apiService = ApiService();

  // Get all complaints
  Future<List<Complaint>> getComplaints({
    String? status,
    String? category,
    String? submittedBy,
    int page = 1,
    int limit = 10,
  }) async {
    try {
      final queryParams = <String, String>{};
      if (status != null) queryParams['status'] = status;
      if (category != null) queryParams['category'] = category;
      if (submittedBy != null) queryParams['submittedBy'] = submittedBy;
      queryParams['page'] = page.toString();
      queryParams['limit'] = limit.toString();

      final response = await _apiService.get(
        '/complaints',
        queryParams: queryParams,
      );

      if (response['success'] == true) {
        final List<dynamic> data = response['data'] ?? [];
        return data.map((json) => Complaint.fromJson(json)).toList();
      }

      throw Exception('Failed to fetch complaints');
    } catch (e) {
      throw Exception('Get complaints error: $e');
    }
  }

  // Get single complaint by ID
  Future<Complaint> getComplaint(String id) async {
    try {
      final response = await _apiService.get('/complaints/$id');

      if (response['success'] == true) {
        return Complaint.fromJson(response['data']);
      }

      throw Exception('Complaint not found');
    } catch (e) {
      throw Exception('Get complaint error: $e');
    }
  }

  // Get complaint by complaint ID (for public tracking)
  Future<Complaint> getComplaintByComplaintId(String complaintId) async {
    try {
      final response = await _apiService.get(
        '/complaints/complaint-id/$complaintId',
        includeAuth: false, // Public endpoint
      );

      if (response['success'] == true) {
        return Complaint.fromJson(response['data']);
      }

      throw Exception('Complaint not found');
    } catch (e) {
      throw Exception('Get complaint error: $e');
    }
  }

  // Create complaint
  Future<Complaint> createComplaint({
    required String title,
    required String description,
    required String category,
    required String address,
    double? latitude,
    double? longitude,
    String priority = 'Medium',
  }) async {
    try {
      final response = await _apiService.post(
        '/complaints',
        {
          'title': title,
          'description': description,
          'category': category,
          'location': {
            'address': address,
            'coordinates': {
              'latitude': latitude,
              'longitude': longitude,
            },
          },
          'priority': priority,
        },
      );

      if (response['success'] == true) {
        return Complaint.fromJson(response['data']);
      }

      throw Exception('Failed to create complaint');
    } catch (e) {
      throw Exception('Create complaint error: $e');
    }
  }

  // Update complaint
  Future<Complaint> updateComplaint(
    String id,
    Map<String, dynamic> data,
  ) async {
    try {
      final response = await _apiService.put(
        '/complaints/$id',
        data,
      );

      if (response['success'] == true) {
        return Complaint.fromJson(response['data']);
      }

      throw Exception('Failed to update complaint');
    } catch (e) {
      throw Exception('Update complaint error: $e');
    }
  }

  // Delete complaint
  Future<void> deleteComplaint(String id) async {
    try {
      final response = await _apiService.delete('/complaints/$id');

      if (response['success'] != true) {
        throw Exception('Failed to delete complaint');
      }
    } catch (e) {
      throw Exception('Delete complaint error: $e');
    }
  }

  // Add comment to complaint
  Future<Complaint> addComment(String complaintId, String text) async {
    try {
      final response = await _apiService.post(
        '/complaints/$complaintId/comments',
        {'text': text},
      );

      if (response['success'] == true) {
        return Complaint.fromJson(response['data']);
      }

      throw Exception('Failed to add comment');
    } catch (e) {
      throw Exception('Add comment error: $e');
    }
  }
}

File: frontend/src/services/complaintService.js
/**
 * complaintService.js - Complaint API Service
 * 
 * This file contains all API calls related to complaints:
 * - Get all complaints (with filters)
 * - Get single complaint
 * - Create complaint
 * - Update complaint
 * - Delete complaint
 * - Add comment
 * - Update status
 * - Track complaint by complaintId
 * 
 * Each function makes HTTP request and returns data
 * Used by: ComplaintContext, Complaint pages
 */

import api from './api'

/**
 * Get all complaints with optional filters
 * @param {Object} filters - { status, category, priority }
 * @returns {Promise} Array of complaints
 */
export const getComplaints = async (filters = {}) => {
  const params = new URLSearchParams()

  // Add filters as query parameters
  if (filters.status) params.append('status', filters.status)
  if (filters.category) params.append('category', filters.category)
  if (filters.priority) params.append('priority', filters.priority)

  const queryString = params.toString()
  const url = queryString ? `/complaints?${queryString}` : '/complaints'

  // Return the full axios response (so callers can access response.data.data)
  const response = await api.get(url)
  return response
}

/**
 * Get single complaint by ID
 * @param {String} id - Complaint ID
 * @returns {Promise} Complaint data
 */
export const getComplaint = async (id) => {
  const response = await api.get(`/complaints/${id}`)
  return response.data
}

/**
 * Get complaint by complaintId (for public tracking)
 * @param {String} complaintId - Complaint ID (e.g., COMP-20240115-12345)
 * @returns {Promise} Complaint data
 */
export const getComplaintByComplaintId = async (complaintId) => {
  const response = await api.get(`/complaints/complaint-id/${complaintId}`)
  return response.data
}

/**
 * Create new complaint
 * @param {Object} complaintData - { title, description, category, location, priority, imageUrl }
 * @returns {Promise} Created complaint data
 */
export const createComplaint = async (complaintData) => {
  const response = await api.post('/complaints', complaintData)
  return response.data
}

/**
 * Update complaint
 * @param {String} id - Complaint ID
 * @param {Object} updateData - Fields to update
 * @returns {Promise} Updated complaint data
 */
export const updateComplaint = async (id, updateData) => {
  const response = await api.put(`/complaints/${id}`, updateData)
  return response.data
}

/**
 * Delete complaint (Admin only)
 * @param {String} id - Complaint ID
 * @returns {Promise} Success message
 */
export const deleteComplaint = async (id) => {
  const response = await api.delete(`/complaints/${id}`)
  return response.data
}

/**
 * Add comment to complaint
 * @param {String} id - Complaint ID
 * @param {String} text - Comment text
 * @returns {Promise} Updated complaint with new comment
 */
export const addComment = async (id, text) => {
  const response = await api.post(`/complaints/${id}/comments`, { text })
  return response.data
}

/**
 * Update complaint status (Officer/Admin only)
 * @param {String} id - Complaint ID
 * @param {String} status - New status
 * @param {String} notes - Optional notes
 * @returns {Promise} Updated complaint data
 */
export const updateStatus = async (id, status, notes = '') => {
  const response = await api.put(`/complaints/${id}/status`, { status, notes })
  return response.data
}


export const assignComplaint = (id, officerId) => {
  return api.put(`/complaints/${id}/assign`, { officerId })
}
